qmp/socket.go

```diff
@@ -104,7 +104,9 @@ func (mon *SocketMonitor) Disconnect() error {
 	atomic.StoreInt32(mon.listeners, 0)
 	err := mon.c.Close()
 
-	for range mon.stream {
+	if mon.stream != nil {
```

Comment:
Its not safe to call Disconnect unless Connect is complete, and is only "safer" with this change,  it seems to me, because it has a  https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use race.

What is your use case? What are you hoping to happen when calling Disconnect when the Connect is in progress? With this change, can't the Disconnect succeed, but since the Connect is ongoing, but then subsequently the Connect can succeed too, leaving the SocketMonitor connected?

I suspect that what is really wanted here is a context.Context argument to Connect(), so that the connection can be canceled or have a timeout.

Reply:
The ill effect will be that after calling Disconnect()... the monitor will get connected. Seems odd to me.

This doesn't break our use of the package, so in that sense, I'm ok with merging it.

It seems to me that if robustness for your use case is required, you'd be better off fixing the underlying bug:

> Basically we are seeing qo-qmp Connect() function block when the qemu process's qmp socket isn't responding (if the qemu process has gone bad for some reason).

and implementing ctx cancel. But that's your call.

Reply:
btw, when calling Disconnect() because Connect() is taking longer than desired, doesn't the goro that is blocked calling Connect() stay around forever, leaking a goro?

Reply:
Thanks!

So because `Disconnect()` calls `mon.c.Close()` first, this will then cause `Connect()` to return an error if it was previously stuck reading/writing to the qmp socket in this section:

https://github.com/digitalocean/go-qemu/blob/f911ede0006740c506532206a6959391f4d98776/qmp/socket.go#L123-L147

So the go routine calling `Connect()` will be cleaned up at that point.

If somehow between calling `Connect()`, it does successfully read/write the preamble to the qmp socket and starts `mon.listen`, but doesn't return in time for `mon.stream` to be initialised by the time `Disconnect()` is called, the `mon.c.Close()` will still be called, which will then cause `mon.listen` go routine to end because it is reading from the socket https://github.com/digitalocean/go-qemu/blob/f911ede0006740c506532206a6959391f4d98776/qmp/socket.go#L176, this will then cause the initialized  `mon.stream` and `mon.events` channels to be closed.

So I think this should be safe to merge and won't cause go routines to leak.

Reply:
@sam-github are you happy to merge this please? Thanks